src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+--    <...>
+-- }
+local function message(message_name, message_def)
+    local field_by_name = {}
+    local field_by_id = {}
+    for field_name, def in pairs(message_def) do
+        local field_type = def[1]
+        local field_id = def[2]
+        if field_by_id[field_id] ~= nil then
+            error(('Id %d in field %q was already used'):format(field_id,
+                field_name))
+        end
+        if field_id < min_field_id or field_id > max_field_id then
+            error(('Id %d in field %q is out of range [%d; %d]'):format(
+                field_id, field_name, min_field_id, max_field_id))
+        end
+        if field_id >= reserved_field_id_min and
+           field_id <= reserved_field_id_max then
+           error(('Id %d in field %q is in reserved ' ..
+               'id range [%d, %d]'):format(field_id, field_name,
+               reserved_field_id_min, reserved_field_id_max))
+        end
+        local field_def = {
+            type = field_type,
+            name = field_name,
+            id = field_id,
+        }
+        field_by_name[field_name] = field_def
+        field_by_id[field_id] = field_def
+    end
+    return {
+        type = 'message',
+        name = message_name,
+        field_by_name = field_by_name,
+        field_by_id = field_by_id
+    }
+end
+
+-- Create a enum object suitable to pass into
+-- the protobuf.protocol function.
+--
+-- Accepts a name of an enum and an enum definition
+-- in the following format.
+--
+-- enum_def = {
+--     <value_name> = <value_id>,
+--     <...>
+-- }
+local function enum(enum_name, enum_def)
+    local value_by_name = {}
+    local value_by_id = {}
+    for value_name, value_id in pairs(enum_def) do
+        if value_by_id[value_id] ~= nil then
+            error(('Double definition of enum field %q by %d'):format(
+                value_name, value_id))
+        end
+        value_by_name[value_name] = value_id
+        value_by_id[value_id] = value_name
+    end
+    return {
+        type = 'enum',
+        name = enum_name,
+        value_by_name = value_by_name,
+        value_by_id = value_by_id,
+    }
+end
+
+-- Create a protocol object that stores message
+-- data needed for encoding.
+--
+-- Accepts protocol definition using protobuf.message
+-- and protobuf.enum functions as in example.
+--
+-- protocol_def = {
+--     protocol.message(<message_name>, <message_def>),
+--     protocol.enum(<enumc_name>, <enum_def>),
+--     <...>
+-- }
+local function protocol(protocol_def)
+    local return_protocol = {}
+    local declarations = {}
+    for _, def in ipairs(protocol_def) do
+        local field_name = def['name']
+        local field_type = def['type']
+        if declarations[field_name] then
+            error(('Double definition of name %q'):format(field_name))
+        end
+        if field_type == 'message' then
+            for _, val in ipairs(def['field_by_id']) do
+                local val_type = val['type']
```

Comment:
We shouldn't traverse the `field_by_id` mapping using `ipairs()`, because the IDs are not necessary starts from 1 and goes to some max ID without any holes.

```lua
tarantool> for k, v in ipairs({1, nil, 3}) do print(k, v) end
1	1
---
...
```

(

BTW, I would highlight the consistent terminology question here once again. `val` is a message field definition. The same name should be used for such an object consistently across the code. `field_def` is already used in the `protobuf.message` code, so I would pick it up.

Similarly, `field_def.type` (it is `val_type` now) should either be used directly as `field_def.type` or assigned to a same named variable consistently across the code.

Reply:
Fixed